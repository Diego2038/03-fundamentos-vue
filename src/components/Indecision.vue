<template>
  <h1>Indecision app</h1>
  <img :src="img" alt="bg" v-if="img">
  <div class="bg-dark"></div>

  <div class="indecision-container">

    <input type="text" placeholder="Hazme una pregunta" v-model="question" @keypress.enter="printQuestion">
    <p>Recuerda terminar con el signo (?)</p>

    <div>
      <h2 v-if="validarPregunta(question)"> {{question}} </h2>
      <h1 v-if="answer"> {{ traducirRespuesta }} </h1>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Indecision',
  data() {
    return {
      question : "",
      answer : null,  
      img : null
    }
  },
  methods : {
    printQuestion(){
      console.log( this.question )
    },
    async getAnswer(){ 
      try {
        this.answer = 'Pensando...' 
        const {answer, image} = await fetch('https://yesno.wtf/api')
         .then( r => r.json())  
       this.img = image
        this.answer = answer 
        //console.log( data ) 
    } catch( err ){
        console.log( 'Indecision Component:', err )
        this.answer = 'No se pudo cargar del API'
        this.img = null
    }
    },
    validarPregunta( palabra ) {
      //return palabra.slice( palabra.length ) === '?'
      return palabra.includes('?')
    }
  },
  watch : {
    question( value, oldValue ){
      console.log({ value })
      if ( !value.includes('?')) return
      // TODO: Realizar petición
      
      this.getAnswer()
    }
  },
  computed : {
    traducirRespuesta () {
      if ( this.answer === 'yes') return 'Sí'
      else if ( this.answer === 'no' ) return 'Nope'
      else if ( this.answer === 'maybe' ) return 'Podría ser'
      else return null
    }
  }
}
</script>

<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>