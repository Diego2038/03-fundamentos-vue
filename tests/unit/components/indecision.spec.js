import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision"


describe("Prueba componentes Indecision Vue", () => {

  let wrapper;
  let clgSpy;

  // Para crear una nueva función en el objeto global. Con esto se "simula" que se tiene fetch, ya que jest se ejecuta con Node, y Node no tiene implementado fetch (sólo los navegadores lo tienen)
  global.fetch = jest.fn( () => Promise.resolve({
    json: () => Promise.resolve({
      // Aquí simulas la devolución del API
      answer : 'yes',
      forced: false,
      image: "https://yesno.wtf/assets/yes/2.gif"
    })
  })) 

  beforeEach( () => {
    wrapper = shallowMount( Indecision )
    clgSpy = jest.spyOn( console, 'log' ) // Es un espía que está pendiente de la función console (no puede ser algo diferente a una función  ), y la función log

    jest.clearAllMocks() // Para limpiar todos los mocks en cada prueba
  })

  test('Hacer match con el snapshot', () => {
    // console.log( wrapper.html() )
    expect( wrapper.html() ).toMatchSnapshot(); // Con esto toda la estructura del HTML queda igual, al menos que ejecutes -r para actualizar

  })

  // Y no debe de ejecutar la función getAnswer
  test('escribir en el input no debe de disparar nada (console.log)', async () => {

    //console.log( wrapper.vm ) // OJO, con vm puedes. vm es la instancia de Vue propiamente
    const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer')

    const input = wrapper.find('input')
    await input.setValue('Hola mundo')
    expect( clgSpy ).toHaveBeenCalled() // ser llamado
    // expect( clgSpy ).toHaveBeenCalledTimes(2) // Ser llamado x cantidad de veces

    expect( getAnswerSpy ).not.toHaveBeenCalled()
  })

  test('escribir el símbolo de "?" debe de disparar el getAnswer', async () => {

    const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer')

    const input = wrapper.find('input')
    await input.setValue('pregunta x?')

    expect( clgSpy ).toHaveBeenCalled()
    expect( getAnswerSpy ).toHaveBeenCalledTimes(1)
  })

  test('pruebas en getAnswer', async () => {
    await wrapper.vm.getAnswer() // Nota que el mock ya hace el llamado por defecto.
    // Así no se agota la cuota del llamado de la API (en caso de ser necesario)

    // console.log( wrapper.vm.img )
    // console.log( wrapper.vm.answer )

    const img = wrapper.find('img')

    expect( img.exists() ).toBeTruthy()
    expect( wrapper.vm.img ).toBe("https://yesno.wtf/assets/yes/2.gif")
    expect( wrapper.vm.answer ).toBe('yes')
  })

  test('pruebas en getAnswer - Fallo en la API', async () => {

    fetch.mockImplementationOnce( () => Promise.reject('Api is down'))
    await wrapper.vm.getAnswer()  
    

    const img = wrapper.find('img')
    expect( img.exists() ).toBeFalsy()
    expect( wrapper.vm.answer ).toBe('No se pudo cargar del API')
  })

})