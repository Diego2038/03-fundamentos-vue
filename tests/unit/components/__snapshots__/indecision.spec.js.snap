// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba componentes Indecision Vue Hacer match con el snapshot 1`] = `
<h1>Indecision app</h1>
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con el signo (?)</p>
  <div>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
